pkg/boot/fit/vfit.go

```diff
@@ -0,0 +1,211 @@
+// Copyright 2021 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+//
+// Performs signature checks on FDT images.
+// Currently supports PGP and raw PKCS1v15 RSA signatures.
+//
+// Expected FDT Format:
+//  Node: images
+//  Node: image_name
+//   P: data
+//   Node: signature*
+//    P: value
+//    P: algo          (ex. 'sha256,rsa4096', 'pgp')
+//    P: signer-name   (Optional)
+//    P: key-name-hint (Optional)
+//
+// vfit aims to be TOCTTOU-safe by reading files into memory before verifying.
+package fit
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/rsa"
+	"fmt"
+	"strings"
+	"unicode"
+
+	"github.com/u-root/u-root/pkg/dt"
+	"github.com/u-root/u-root/pkg/vfile"
+	"golang.org/x/crypto/openpgp"
+)
+
+var (
+	algs = map[string]crypto.Hash{
+		"MD4":       crypto.MD4,
+		"MD5":       crypto.MD5,
+		"SHA1":      crypto.SHA1,
+		"SHA224":    crypto.SHA224,
+		"SHA256":    crypto.SHA256,
+		"SHA384":    crypto.SHA384,
+		"SHA512":    crypto.SHA512,
+		"RIPEMD160": crypto.RIPEMD160,
+		"SHA3_224":  crypto.SHA3_224,
+		"SHA3_256":  crypto.SHA3_256,
+		"SHA3_384":  crypto.SHA3_384,
+		"SHA3_512":  crypto.SHA3_512,
+	}
+)
+
+type Signature interface {
+	fmt.Stringer
+	// Warning: If the signature does not exist or does not match the keyring,
+	// both the file and a signature error will be returned.
+	Verify([]byte, openpgp.KeyRing) (*bytes.Reader, error)
+}
+
+type PGPSignature struct {
+	name  string // Name of signature Node
+	value []byte
+	// Optional description fields
+	signer string
+	hint   string
+}
+
+type RSASignature struct {
+	name  string // Name of signature Node
+	hash  crypto.Hash
+	value []byte
+	// Optional description fields
+	signer string
+	hint   string
+}
+
+func (s PGPSignature) String() string {
+	return fmt.Sprintf("PGP Signature - name: %s, signer: '%s', hint: '%s'", s.name, s.signer, s.hint)
+}
+
+// Warning: If the signature does not exist or does not match the keyring,
+// both the file and a signature error will be returned.
```

Comment:
I feel a little weird about this pattern, but I acknowledge that a similar one is already employed for vfile. As a relative newcomer to reviewing u-root changes (really, I'm just trying to be thorough filling in for @hugelgupf 😅), I'll offer this commentary but not block the review.

From a caller ergonomics perspective, I disagree with returning both the file and the signature error in this case. I think the fact that there is a warning about unintuitive behavior is a bit of a smell. (vfile has this issue too, and I would offer the same comment there; one difference I'd acknowledge is that those functions take a path and return a `vfile.File` so it's kind of pulling double duty as a replacement to `os.Open`)

My problem with this pattern is that a caller who doesn't read the warning may write something like:

```go
var myStuff []byte // These came from somewhere, reading a file or something
rdr, _ := someSignature.Verify(myStuff)
if rdr != nil {
    // I guess verification succeeded
    doSomethingWith(rdr)
}
```

If there is no signature or the signature was invalid, and the caller still wants to use the thing-that-was-maybe-signed, I'd say they should use their copy of it (which they already have totally in memory as opposed to only a path to it). There is no TOCTTOU concern if the 'C' is optional 😁.

I think you may be able to remove potential footguns if you make these functions accept an io.Reader and return an io.ReadSeeker. Callers can use the logic above (if they don't care what the error was) or call with a more clear flow:

```go
var myStuff []byte // These came from somewhere, reading a file or something
var rdr io.Reader
rdr = bytes.NewReader(myStuff)
if verified, err := someSignature.Verify(rdr); err == nil {
    // Use the TOCTTOU-proof thing that was verified
    rdr = verified
}
doSomethingWith(rdr)
```

Reply:
There is always a potential for misuse, but I overall agree it doesn't feel very golang. 
However, I'll let @hugelgupf return before changing the established interface here.

Reply:
SG. I'll resolve this here and create an email in case he doesn't dig through this PR when he returns.